// react-relay-network-modern/node8/__mocks__/mockReq.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.mockFormDataReq = mockFormDataReq;
exports.mockMutationReq = mockMutationReq;
exports.mockReq = mockReq;
exports.mockReqWithFiles = mockReqWithFiles;
exports.mockReqWithSize = mockReqWithSize;

function ownKeys(object, enumerableOnly) { var keys = Object.keys(object); if (Object.getOwnPropertySymbols) { var symbols = Object.getOwnPropertySymbols(object); if (enumerableOnly) { symbols = symbols.filter(function (sym) { return Object.getOwnPropertyDescriptor(object, sym).enumerable; }); } keys.push.apply(keys, symbols); } return keys; }

function _objectSpread(target) { for (var i = 1; i < arguments.length; i++) { var source = arguments[i] != null ? arguments[i] : {}; if (i % 2) { ownKeys(Object(source), true).forEach(function (key) { _defineProperty(target, key, source[key]); }); } else if (Object.getOwnPropertyDescriptors) { Object.defineProperties(target, Object.getOwnPropertyDescriptors(source)); } else { ownKeys(Object(source)).forEach(function (key) { Object.defineProperty(target, key, Object.getOwnPropertyDescriptor(source, key)); }); } } return target; }

function _defineProperty(obj, key, value) { if (key in obj) { Object.defineProperty(obj, key, { value: value, enumerable: true, configurable: true, writable: true }); } else { obj[key] = value; } return obj; }

/* eslint-disable import/prefer-default-export, no-param-reassign */
class MockReq {
  constructor(reqid, reqData = {}) {
    this.reqid = reqid || Math.random().toString();
    this.reqData = reqData;
  }

  getID() {
    return this.reqid;
  }

  getQueryString() {
    return this.reqData.query || '';
  }

  getDebugName() {
    return `debugname${this.reqid}`;
  }

  getVariables() {
    return this.reqData.variables || {};
  }

  getFiles() {
    return this.reqData.files;
  }

  reject(err) {
    this.error = err;
  }

  resolve(resp) {
    this.payload = resp;
  }

  execute(rnl) {
    const text = this.getQueryString() || '';
    const operationKind = text.startsWith('mutation') ? 'mutation' : 'query';
    const operation = {
      id: this.getID(),
      text,
      operationKind
    };
    const variables = this.getVariables() || {};
    const cacheConfig = this.reqData.cacheConfig || {};
    const uploadables = this.getFiles();
    const res = rnl.fetchFn(operation, variables, cacheConfig, uploadables);
    const promise = new Promise((resolve, reject) => {
      res.subscribe({
        complete: () => {},
        error: error => reject(error),
        next: value => resolve(value)
      });
    }); // avoid unhandled rejection in tests

    promise.catch(() => {}); // but allow to read rejected response

    return promise;
  }

}

function mockReq(reqid, data) {
  return new MockReq(reqid ? reqid.toString() : undefined, data);
}

function mockMutationReq(reqid, data) {
  return new MockReq(reqid ? reqid.toString() : undefined, _objectSpread({
    query: 'mutation {}'
  }, data));
}

function mockFormDataReq(reqid, data) {
  return new MockReq(reqid ? reqid.toString() : undefined, _objectSpread({
    files: {
      file1: 'data'
    }
  }, data));
}

function mockReqWithSize(reqid, size) {
  return mockReq(reqid, {
    query: `{${'x'.repeat(size)}}`
  });
}

function mockReqWithFiles(reqid) {
  return mockReq(reqid, {
    files: {
      file1: 'data',
      file2: 'data'
    }
  });
}
