// react-relay-network-modern/lib/middlewares/logger.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports["default"] = loggerMiddleware;

var _RelayRequest = _interopRequireDefault(require("../RelayRequest"));

var _RelayRequestBatch = _interopRequireDefault(require("../RelayRequestBatch"));

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { "default": obj }; }

function ownKeys(object, enumerableOnly) { var keys = Object.keys(object); if (Object.getOwnPropertySymbols) { var symbols = Object.getOwnPropertySymbols(object); if (enumerableOnly) { symbols = symbols.filter(function (sym) { return Object.getOwnPropertyDescriptor(object, sym).enumerable; }); } keys.push.apply(keys, symbols); } return keys; }

function _objectSpread(target) { for (var i = 1; i < arguments.length; i++) { var source = arguments[i] != null ? arguments[i] : {}; if (i % 2) { ownKeys(Object(source), true).forEach(function (key) { _defineProperty(target, key, source[key]); }); } else if (Object.getOwnPropertyDescriptors) { Object.defineProperties(target, Object.getOwnPropertyDescriptors(source)); } else { ownKeys(Object(source)).forEach(function (key) { Object.defineProperty(target, key, Object.getOwnPropertyDescriptor(source, key)); }); } } return target; }

function _defineProperty(obj, key, value) { if (key in obj) { Object.defineProperty(obj, key, { value: value, enumerable: true, configurable: true, writable: true }); } else { obj[key] = value; } return obj; }

function loggerMiddleware(opts) {
  var logger = opts && opts.logger || console.log.bind(console, '[RELAY-NETWORK]');
  return function (next) {
    return function (req) {
      var start = new Date().getTime();
      logger("Run ".concat(req.getID()), req);
      return next(req).then(function (res) {
        var end = new Date().getTime();
        var queryId;
        var queryData;

        if (req instanceof _RelayRequest["default"]) {
          queryId = req.getID();
          queryData = {
            query: req.getQueryString(),
            variables: req.getVariables()
          };
        } else if (req instanceof _RelayRequestBatch["default"]) {
          queryId = req.getID();
          queryData = {
            requestList: req.requests,
            responseList: res.json
          };
        } else {
          queryId = 'CustomRequest';
          queryData = {};
        }

        logger("Done ".concat(queryId, " in ").concat(end - start, "ms"), _objectSpread(_objectSpread({}, queryData), {}, {
          req: req,
          res: res
        }));

        if (res.status !== 200) {
          logger("Status ".concat(res.status, ": ").concat(res.statusText || '', " for ").concat(queryId));
        }

        return res;
      }, function (error) {
        if (error && error.name && error.name === 'AbortError') {
          logger("Cancelled ".concat(req.getID()));
        }

        throw error;
      });
    };
  };
}
